Collection editing revamp

emma-sg opened the pull request:
Closes #2328 
Closes #2338

Tentatively ready for review...? This is a lot of changes, I'll split them out into sections below.

# Changes

This is a fairly broad pass over the collection editing and sharing experience, focusing primarily on being able to edit collections from anywhere you might want to edit them (i.e. not just from within their detail page).

As part of these changes, we've removed the previous "metadata" editing dialog, and moved name, summary, thumbnail selection (detailed further below), and visibility into a new dialog that can be access from the org dashboard and the collections list, in addition to the collection detail pages.

## Back-end changes

We've added a new `thumbnailSource` field to the Collection model, which preserves the page selected to be collection's thumbnail image. As such, we've separated thumbnail selection from initial replay view selection, though you can still use the initial replay view selection dialog to set the thumbnail if you wish. This field is set via three additional optional query parameters when uploading a thumbnail, and is returned to the client as part of the full collection object.

## Front-end changes

This PR has changed quite a bit since the start of it & the call & issue that prompted it, I'd initially taken the route of trying to combine all collection editing interfaces into a single dialog, but along the way we figured out that doing so would require either stacked dialogs, which we decided against, or merging the "select archived items" dialog into the editing dialog, which would be a lot more unwieldy state logic to deal with, and potentially confusing options too coming from different places.

There's a few things left to do still:
- [x] ~~The current share dialog is still used, and still has visibility & thumbnail selection. This should be removed, and turned into just a dialog with url & embed information.~~ 
  *Completed in 8eb179d*
- [x] The "edit collection" button should maybe be worded differently — or maybe removed entirely? The dialog it opens can be accessed from the inline edit button on the title/summary already, and from sharing settings
- ~~We might not get to it this release, but it'd be great to have a button for setting the currently visible page in the replay to the initial replay view~~
  - I am de-scoping this.  The current button works well enough for me to be happy releasing it.  -HW
- [ ] Various issues when selecting thumbnails — we could exclude pages without thumbnails by excluding uploaded pages, it sounds like? https://github.com/webrecorder/browsertrix/pull/2332#issuecomment-2625647229
  *Sort of handled 32078be with better error states & messaging, but could be improved on further*
  - [x] Checkmark on page selection input is a little misleading when selecting a page without a thumbnail — checkmark should change to a warning icon
  - [x] Focus the URL field when users click the empty thumbnail in the presentation window
  - [x] For some collections, selecting a page doesn't even show the "no preview" info in the preview — logic there isn't quite right
  - [x] Sometimes selecting a page doesn't cause the thumbnail preview to update at all
-  ~~Rename "about" and "about this collection" to something having more to do specifically with the description~~
  - Not doing this!  -HW, ESG
- [x] Rename "metadata" to something else (chose "details")
- [x] Ensure language around "set initial view" is aligned.
- [x] Add "Update starting URL" checkbox when updating thumbnail source
- [x] Fix download button (api name changed and needed to be fixed on the front end)
  - [x] Remove public download button from private collection view
- [x] Update "Collection Settings" button to "Edit Collection Settings" icon button
- [x] Add "Set Initial View" and "Edit Description" to action menu
- [x] Remove share button visible when collection is private
- [ ] Have dashboard collection grid show all collections (maybe tab for public only?) — started on this in #2379!
- [x] Move globe icon on org dashboard back down to Public Collections section
- [x] Rename "Public Gallery" -> "Public Collections Gallery" @SuaYoo

emma-sg committed:
wip

emma-sg committed:
name/summary/description editor (part 1)

emma-sg committed:
share dialog

emma-sg committed:
track dirty state

emma-sg committed:
tweaks

emma-sg committed:
more little fixes

emma-sg committed:
wip uploads

emma-sg committed:
allow aborting uploads

emma-sg committed:
merge homepage editing & updates into editing dialog

emma-sg committed:
refactor logic into separate files

emma-sg committed:
more refactors

emma-sg committed:
allow opening to a specific tab

emma-sg committed:
tun off page url autocomplete (it has its own)

emma-sg committed:
fix thumbnail & snapshot state passing

emma-sg committed:
improve editing ux

emma-sg committed:
ux tweaks

emma-sg committed:
add edit icon button to header

emma-sg committed:
allow editing collection from collection grid

emma-sg committed:
use edit collection in collection list, and lots of small fixes

emma-sg committed:
Merge remote-tracking branch 'origin/main' into frontend-collection-editing-dialog

SuaYoo committed:
feat: Update references to org public profile -> gallery (#2330)

- Renames public URL prefix to `explore`
- Updates org settings sections
- Removes or renames references to "org profile"

---------

Co-authored-by: Ilya Kreymer <ikreymer@gmail.com>
Co-authored-by: Henry Wilkinson <henry@wilkinson.graphics>

SuaYoo committed:
feat: Update references to org public profile -> gallery (#2330)

- Renames public URL prefix to `explore`
- Updates org settings sections
- Removes or renames references to "org profile"

---------

Co-authored-by: Ilya Kreymer <ikreymer@gmail.com>
Co-authored-by: Henry Wilkinson <henry@wilkinson.graphics>

emma-sg committed:
wip remove homepage & move thumbnail selector to general section

emma-sg committed:
wip

Shrinks99 committed:
Remove duplicate dialogue actions from collections

Shrinks99 committed:
Remove duplicate dialogue actions from collections

Shrinks99 committed:
Add remaining icon buttons

Still needs logic for changing icons based on view

Shrinks99 committed:
Add remaining icon buttons

Still needs logic for changing icons based on view

Shrinks99 committed:
Remove share item from menu

Shrinks99 committed:
Remove button group

Shrinks99 committed:
Move edit button for logged in users

Shrinks99 committed:
fix eslint error

non nullish value

Shrinks99 committed:
Add dashboard link

Shrinks99 committed:
Simplify collection grid action dropdown → edit button

TODO: Actually hook up this button to a dialogue @emma-sg

Shrinks99 committed:
Button casing & consistency fixes

Shrinks99 committed:
Align pageHeader actions to bottom of container

emma-sg committed:
get most aspects of dialog working!

emma-sg committed:
get thumbnail selection working

emma-sg committed:
first pass at backend update to allow thumbnail src to be saved

Shrinks99 committed:
Adds org description and url to dashboard

Shrinks99 committed:
Re-remove duped menu item

Shrinks99 committed:
Restore pencil and fix primary action buttons

emma-sg committed:
redo thumbnail source as separate field on collection model

emma-sg committed:
fix lint issue

emma-sg committed:
fix missed type change

emma-sg committed:
hook up edit button to edit dialog

Shrinks99 committed:
Remove view public button from share-collection

Shrinks99 committed:
wip: broken logic for moving public collection view to top right

emma-sg committed:
refactor shareLink into separate fn

Shrinks99 committed:
Remove view public from share-collection controls

emma-sg committed:
pass thumbnail source into thumbnail source update fn

emma-sg committed:
fix search param names

Shrinks99 committed:
temp: Move collection gallery link to icon button

Better than where it was for now!

emma-sg committed:
wip

emma-sg committed:
lint & format

emma-sg committed:
Merge pull request #2344 from webrecorder/frontend-collection-control-constriction

Button reworks and navigation updates

Shrinks99 committed:
update button text

is it the same?  not quite.... is it done?  yes!

Shrinks99 committed:
Removes dropdown, moves controls to icon buttons

Shrinks99 committed:
make explicit that this code isn't doing anything

emma-sg committed:
swap summary textarea for input

emma-sg committed:
Merge branch 'frontend-collection-editing-dialog' of https://github.com/webrecorder/browsertrix into frontend-collection-editing-dialog

emma-sg committed:
clean up clipboard button value

emma-sg committed:
fix layout shift issues on collection detail page

Shrinks99 committed:
dash → en dash on collection thumbnail badge

emma-sg committed:
fix a whole bunch of small-ish issues

emma-sg committed:
Merge branch 'frontend-collection-editing-dialog' of https://github.com/webrecorder/browsertrix into frontend-collection-editing-dialog

emma-sg committed:
polish - more consistent spacing & wording, small bugfixes, etc

emma-sg committed:
reset all things when resetting dialog

emma-sg committed:
tweak snapshot preview help text

emma-sg committed:
simplify logic for allowing selecting page thumbnail

emma-sg committed:
remove console logs

emma-sg committed:
Merge branch 'main' into frontend-collection-editing-dialog

emma-sg committed:
change collection metadata dialog into just collection create dialog

emma-sg committed:
clean up missed extra attr

emma-sg committed:
remove extra `/` in api.upload method

emma-sg committed:
fix missed renames

emma-sg committed:
remove name field customization

emma-sg committed:
exit early out of blob task if snapshot fields aren't defined

emma-sg committed:
fix missing type conversions

emma-sg committed:
tweak text rendering

emma-sg committed:
fix some edge cases

emma-sg committed:
allow resetting thumbnail source

ikreymer commented:
Some initial feedback:
- Since uploads don't have screenshots currently, we could exclude all pages from uploads, though we could support them in the future. Maybe we just say 'Sorry, this page doesn't have a preview' rather than 'couldn't load preview' to be more clear?
- Find the disabled 'x' a bit confusing, I feel like if I click x I want to discard the changes - that seems like it should just discard and close.

ikreymer commented:
- Looks like there's still a standalone share button which shows the old share dialog on a new collection. This is redundant with the Sharing tab on Edit Collection: 
<img width="1305" alt="Screenshot 2025-01-30 at 1 43 31 PM" src="https://github.com/user-attachments/assets/d41c86f2-400c-4ecd-918a-1976bbbbc910" />

emma-sg committed:
fix a couple issues with state management

emma-sg committed:
clean up sharing dialog

emma-sg committed:
allow closing the dialog with the top right x icon, and fix an issue with initial states

emma-sg commented:
> Find the disabled 'x' a bit confusing

Sure, yeah, that's fair! Re-enabled it in 77a47ec.

> Looks like there's still a standalone share button which shows the old share dialog

Should be addressed in 8eb179d!

> Since uploads don't have screenshots currently, we could exclude all pages from uploads

Is there an easy way to do this? The way thing are currently, I don't think there's a way without switching from querying the db to querying RWP, unless I'm mistaken. It'd be great to do though! I'll take a pass at improving the error/not found messaging in the meantime.

emma-sg committed:
clean up thumbnail selector & add better error handling & informing

emma-sg committed:
Merge branch 'main' into frontend-collection-editing-dialog

Shrinks99 committed:
Changes metadata → details

Shrinks99 committed:
metadata → details (public view)

emma-sg committed:
rename select collection page file

emma-sg committed:
use click handler on button rather than icon

Shrinks99 committed:
Unify initial view naming

Shrinks99 committed:
Edit Collection → Collection Settings

Shrinks99 committed:
Update workflow edit icon button

- Makes this uniform with other in-page edit buttons which use the pencil
- Adds a tooltip

Shrinks99 committed:
Update collection editing icons and tab names

Shrinks99 committed:
Revert one of the gear updates

oops

emma-sg committed:
fix the thing!!! thank you sua

emma-sg committed:
Merge branch 'frontend-collection-editing-dialog' of https://github.com/webrecorder/browsertrix into frontend-collection-editing-dialog

emma-sg committed:
allow thumbnail list to scroll

emma-sg committed:
tweak min widths & add sticky header

emma-sg committed:
fix download url

Shrinks99 committed:
Remove share options from collection list actions

Effectively duped, access through edit settings instead.

- Adds icon and variant state to copy toast

emma-sg committed:
add "set initial view to this page" checkbox to edit dialog

emma-sg committed:
Merge branch 'frontend-collection-editing-dialog' of https://github.com/webrecorder/browsertrix into frontend-collection-editing-dialog

emma-sg committed:
clean up

emma-sg committed:
Merge branch 'main' into frontend-collection-editing-dialog

emma-sg committed:
resolve layout issues

emma-sg committed:
remove download button from private detail view

emma-sg committed:
small bugfixes

emma-sg committed:
don't allow actually selecting page thumbnail until there's a page thumbnail to select

emma-sg committed:
fix a couple subtle page selection indicator bugs

emma-sg committed:
remove a few `console.log`s

emma-sg committed:
fix update ordering

emma-sg committed:
update error to warning rather than error when thumbnail isnt actually selected

emma-sg committed:
remove more `console.log`s

SuaYoo commented on frontend/src/context/view-state.ts:
```diff
@@ -0,0 +1,7 @@
+
+export type ViewStateContext = ViewState | null;
+
+export const viewStateContext = createContext<ViewStateContext>("viewState");
```
Nice!

SuaYoo commented on frontend/src/features/collections/collections-grid.ts:
```diff
@@ -95,26 +105,50 @@ export class CollectionsGrid extends BtrixElement {
+            @sl-after-hide=${() => {
+              this.collectionBeingEdited = null;
+              // TODO propagate an event back up & refresh collections
+              // void this.fetchCollection();
```
Is this TODO still relevant?

emma-sg replied:
Good catch, no, I don't think so — it's handled in the org dashboard with the `btrix-collection-saved` event

SuaYoo commented on frontend/src/features/collections/edit-dialog/general-section.ts:
```diff
@@ -0,0 +1,320 @@
+      type="button"
+      aria-checked=${isSelected}
+      @click=${() => {
+        this.thumbnailSelector?.input?.focus();
```
Seeing an issue where if I click and then "Page URL", or vice versa, the dropdown for "Page URL" doesn't open.

emma-sg replied:
Can reproduce, looking into this

emma-sg committed:
fix uploading logic & pass correct params in initial view selection dialog

SuaYoo commented on frontend/src/features/collections/edit-dialog/general-section.ts:
```diff
@@ -0,0 +1,320 @@
+            this.selectedSnapshot = newSnapshot;
+          }
+
+          void this.checkChanged();
```
If a URL that's not in the collection is typed in, it doesn't trigger the changes check. That behavior seems correct, but then I'm able to close out of the dialog, and when it's reopened the old value persists. We'll probably want to reset the form fields on `sl-after-hide`. Clicking "Cancel" also doesn't reset the Page URL field.

SuaYoo commented on frontend/src/pages/org/collection-detail.ts:
```diff
@@ -409,37 +499,18 @@ export class CollectionDetail extends BtrixElement {
               }
 
-              this.openDialogName = "editStartPage";
+              this.openDialogName = "replaySettings";
```
Is this intended to be for editing the initial view or a repeat of the general edit collection button? If the former, the label will need to be updated. If latter, open dialog name should be `edit`.

emma-sg committed:
changes from code review

Co-authored-by: sua yoo <sua@webrecorder.org>

emma-sg committed:
reset page selector on dialog reset

emma-sg committed:
Merge branch 'main' into frontend-collection-editing-dialog

SuaYoo commented:
Small thing I noticed--can we add a minimum height to the edit dialog, so that the location of navigation tab buttons in the viewport don't shift as much, if at all?

emma-sg committed:
hide collection share buttons when collection is private

SuaYoo committed:
update copy

emma-sg committed:
move "visit public collections gallery" button down to public collections list

emma-sg committed:
Merge branch 'frontend-collection-editing-dialog' of https://github.com/webrecorder/browsertrix into frontend-collection-editing-dialog

emma-sg committed:
add "select initial view" and "edit description" to action menu

emma-sg committed:
change edit button to icon button

emma-sg committed:
ensure "edit collection settings" is used everywhere

emma-sg committed:
update "change org settings" to "edit org settings"

SuaYoo committed:
make summary clickable

emma-sg committed:
fix & add tests for thumbnail upload

emma-sg committed:
Merge branch 'frontend-collection-editing-dialog' of https://github.com/webrecorder/browsertrix into frontend-collection-editing-dialog

emma-sg committed:
make set initial view checkbox not checked by default
this was causing some edge cases where opening the dialog would leave you with unsaved changes immediately on opening, which was confusing

emma-sg committed:
rename "general" to "presentation" to better match tab titles

emma-sg committed:
fix issue where collection caption isn't always null when empty

SuaYoo committed:
handle no pages

emma-sg committed:
add delay to closing combobox in page selection component

emma-sg committed:
Merge branch 'frontend-collection-editing-dialog' of https://github.com/webrecorder/browsertrix into frontend-collection-editing-dialog

emma-sg committed:
fix test

emma-sg committed:
add comment about why timeout in page selector is useful

emma-sg committed:
revert unused change to md editor event

SuaYoo committed:
remove console log

ikreymer reviewed (approved):
backend changes make sense! Should be good to go!
